include/envoy/stream_info/filter_state.h

```diff
@@ -55,15 +56,88 @@ class FilterState {
             (dynamic_cast<const T*>(getDataGeneric(data_name)) != nullptr));
   }
 
+  /**
+   * The addToList, hasListWithName, forEachListItem are variants of the above
+   * functions, that operate on list data. Multiple elements could be added
+   * to the list under the same data_name.
+   * @param data_name the name of the list data being set.
+   * @param data an owning pointer to the data to be appended to the list.
+   * Note that data_names for list elements do not share the same namespace as
+   * the data_names for singleton data objects added through setData. All items
+   * added to the list must be of the same type.
+   */
+  template <typename T>
+  void addToList(absl::string_view data_name, std::unique_ptr<Object>&& data) {
+    const auto* list = getList(data_name);
+    if (list != nullptr) {
+      // Check type of first element in the list
+      const T* cast = dynamic_cast<const T*>(list->at(0).get());
+      if (!cast) {
+        throw EnvoyException(
+            fmt::format("List {} does not conform to the specified type", data_name));
+      }
+    }
+
+    addToListGeneric(data_name, std::move(data));
+  }
+
+  /**
+   * @param data_name the name of the list being probed.
+   * @return Whether a list of the type and name specified exists in the
+   * data store.
+   */
+  template <typename T> bool hasList(absl::string_view data_name) const {
+    const auto* list = getList(data_name);
+    return ((list != nullptr) && (dynamic_cast<const T*>(list->at(0).get()) != nullptr));
+  }
+
+  /**
+   * @param data_name the name of the list data being looked up.
+   * @param operation a lambda function that operates on each element in the list,
+   * if it exists. The iteration will stop if the lambda function returns false or
+   * reaches the end of the list. The iteration order will be the same as the order
+   * in which the elements were added to the list. Note that if the elements in the
+   * list cannot be dynamically type-cast into the requested type, an exception
+   * will be thrown.
+   */
+  template <typename T>
+  void forEachListItem(absl::string_view data_name, std::function<bool(const T&)> op) const {
```

Comment:
Did you consider making this capability via an iterator?

Reply:
Yes.. but the filter state stores unary data and list data in one object. Implementing the iterator interface would be strange in that case. Also, this is doing a hashmap lookup and then doing the iteration/dynamic_cast/break_semantics/etc.. Not sure if all can be covered in teh iterator iterator interface

Reply:
I think it's possible, but I won't push too strongly here, as it would make a more complicated interface.

Reply:
I think that was considered. For context, here are the annals of bike shedding with regard to this topic: https://github.com/envoyproxy/envoy/pull/4588#issuecomment-426605911